src/Agent/NewRelic/Agent/Core/Agent.cs

```diff
@@ -562,19 +567,82 @@ internal void Detach(bool removeAsync, bool removePrimary)
 
             var filteredContextData = new Dictionary<string, object>();
 
-            // first, handle the "includes"
-            if (_configurationService.Configuration.ContextDataInclude.Count() == 0) // empty "include" list, so include everything
+            // Pseudocode
+            // Include/exclude rules:
+            // 1. Exclude wins over include
+            // 2. Can have a '*' wildcard at the end of an include/exclude definition
+            // 3. "more specific" rules win over "less specific".  Example: include=AB, exclude=A*, AB would be included but not ABC or AAA
+
+            // Algorithm:
+            // Create list of rules, sorted by least specific to most specific, with includes first and then excludes within each level of specificity
+            // Specificity is just the length of the string, minus any wildcard character at the end
+            // Apply rules in order. Include rules add attributes to filteredContextData from unfilteredContextData, and exclude rules remove attributes from filteredContextData
+
+            var allRulesInOrder = _orderedCludeRuleList ?? (_orderedCludeRuleList = GetOrderedCludeRuleList()); 
+
+            if (_includeRuleList.Count() == 0) // empty "include" list, so include everything
             {
                 filteredContextData = unfilteredContextData;
             }
-            else
+
+            // Now apply the rules in order
+            // TODO: maybe optimize by just working with keys as lists of strings until it comes time to return the actual data
```

Comment:
Yeah I'm a little concerned about the performance of doing a Where/Concat/ToDictionary for every rule, if there are lot of rules and/or a lot of logging.

Reply:
I feel like if we just do string comparisons on the keys, we might be able to optimize it by caching previous matches somewhere so that the agent doesn't have to do Where/Concat/ToDictionary again.

Reply:
My latest push introduces a cache for include/exclude results, and reverses the logic - instead of going through every rule and applying it to the context data dictionary, it goes through every key/value pair in the context dictionary and looks for a cache result or matching rule.  I think this should be more efficient.